bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+            let msg: Msg = match ciborium::from_reader(
+                &self.read_buf[FRAME_HEADER_SIZE..end],
+            ) {
+                Ok(msg) => {
+                    // Move any remaining bytes to the beginning of the buffer.
+                    self.read_buf.copy_within(end..self.total_read, 0);
+                    self.total_read = self.total_read - end;
+                    msg
+                }
+                Err(e) => {
+                    warn!(
+                        self.log,
+                        "Closing connection: failed to deserialize: {}", e
+                    );
+                    return self.close().await;
+                }
+            };
+            self.last_received_msg = Instant::now();
+            debug!(self.log, "Received {:?}", msg);
+            if let Msg::Fsm(msg) = msg {
+                if let Err(e) = self
+                    .main_tx
+                    .send(ConnToMainMsg {
+                        handle_unique_id: self.unique_id,
+                        msg: ConnToMainMsgInner::Received {
+                            from: self.peer_id.clone(),
+                            msg,
+                        },
+                    })
+                    .await
+                {
+                    warn!(
+                        self.log,
+                        "Failed to send received msg to main task: {:?}", e
+                    );
+                }
+            }
+        }
+    }
+
+    // Send ping messages and check for inactivity timeouts
+    async fn ping(&mut self) -> Result<(), ConnErr> {
+        if Instant::now() - self.last_received_msg > INACTIVITY_TIMEOUT {
+            warn!(self.log, "Closing connection: inactivity timeout",);
+            return self.close().await;
+        }
+        if self.write_queue.len() == MSG_WRITE_QUEUE_CAPACITY {
+            warn!(self.log, "Closing connection: write queue full",);
+            self.close().await
+        } else {
+            match write_framed(&Msg::Ping) {
+                Ok(msg) => {
+                    self.write_queue.push_back(msg);
+                    Ok(())
+                }
+                Err(e) => {
+                    warn!(
+                        self.log,
+                        "Closing connection: Failed to serialize msg: {}", e
+                    );
+                    self.close().await
+                }
+            }
+        }
+    }
+
+    // Close and drop the connection.
+    async fn close(&mut self) -> Result<(), ConnErr> {
+        if let Err(e) = self
+            .main_tx
+            .send(ConnToMainMsg {
+                handle_unique_id: self.unique_id,
+                msg: ConnToMainMsgInner::Disconnected {
+                    peer_id: self.peer_id.clone(),
+                },
+            })
+            .await
+        {
+            warn!(self.log, "Failed to send to main task: {:?}", e);
+        }
+        let _ = self.write_sock.shutdown().await;
```

Comment:
This makes me a little nervous, because it seems possible to leave this task in a broken state.  On the other hand, this function can _only_ return an error, so maybe it's fine (as long as no one ignores the error)?

Reply:
I don't think anything else can be done here. This is also somewhat of a "courtesy call" for the other end of the socket. The socket itself is going to be dropped which should cleanup resources.